src/reliable_text.c

```diff
@@ -0,0 +1,346 @@
+    
+    struct LDPC ldpc;
+} reliable_text_impl_t;
+
+// 6 bit character set for text field use:
+// 0: ASCII null
+// 1-9: ASCII 38-47
+// 10-19: ASCII '0'-'9'
+// 20-46: ASCII 'A'-'Z'
+// 47: ASCII ' '
+static void convert_callsign_to_ota_string_(const char* input, char* output, int maxLength)
+{
+    int outidx = 0;
+    for (size_t index = 0; index < maxLength; index++)
+    {
+        if (input[index] == 0) break;
+        
+        if (input[index] >= 38 && input[index] <= 47)
+        {
+            output[outidx++] = input[index] - 37;
+        }
+        else if (input[index] >= '0' && input[index] <= '9')
+        {
+            output[outidx++] = input[index] - '0' + 10;
+        }
+        else if (input[index] >= 'A' && input[index] <= 'Z')
+        {
+            output[outidx++] = input[index] - 'A' + 20;
+        }
+        else if (input[index] >= 'a' && input[index] <= 'z')
+        {
+            output[outidx++] = toupper(input[index]) - 'A' + 20;
+        }
+    }
+    output[outidx] = 0;
+}
+
+static void convert_ota_string_to_callsign_(const char* input, char* output, int maxLength)
+{
+    int outidx = 0;
+    for (size_t index = 0; index < maxLength; index++)
+    {
+        if (input[index] == 0) break;
+        
+        if (input[index] >= 1 && input[index] <= 9)
+        {
+            output[outidx++] = input[index] + 37;
+        }
+        else if (input[index] >= 10 && input[index] <= 19)
+        {
+            output[outidx++] = input[index] - 10 + '0';
+        }
+        else if (input[index] >= 20 && input[index] <= 46)
+        {
+            output[outidx++] = input[index] - 20 + 'A';
+        }
+    }
+    output[outidx] = 0;
+}
+
+static char calculateCRC8_(char* input, int length)
+{
+    unsigned char generator = 0x1D;
+    unsigned char crc = 0; /* start with 0 so first byte can be 'xored' in */
+
+    while (length > 0)
+    {
+        unsigned char ch = *input++;
+        length--;
+
+        // Ignore 6-bit carriage return and sync characters.
+        if (ch == 63 || ch == 48) continue;
+        
+        crc ^= ch; /* XOR-in the next input byte */
+        
+        for (int i = 0; i < 8; i++)
+        {
+            if ((crc & 0x80) != 0)
+            {
+                crc = (unsigned char)((crc << 1) ^ generator);
+            }
+            else
+            {
+                crc <<= 1;
+            }
+        }
+    }
+
+    return crc;
+}
+
+static int reliable_text_ldpc_decode(reliable_text_impl_t* obj, char* dest, char* src)
+{
+    float incomingData[LDPC_TOTAL_SIZE_BITS];
+    float llr[LDPC_TOTAL_SIZE_BITS];
+    unsigned char output[LDPC_TOTAL_SIZE_BITS];
+    int parityCheckCount = 0;
+    
+    for (int bitIndex = 0; bitIndex < LDPC_TOTAL_SIZE_BITS; bitIndex++)
+    {
+        // Map to value expected by sd_to_llr().
+        int bit = (src[bitIndex / 7] & (1 << (bitIndex % 7))) != 0;
+        incomingData[bitIndex] = 1.0 - 2.0 * bit;
+    }
+    
```

Comment:
This is hard decision decoding, so you will lose much of the power of the LDPC code......